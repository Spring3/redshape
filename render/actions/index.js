import issueActions from './issue.actions';
import timeEntryActions from './timeEntry.actions';
import settingsActions from './settings.actions';

export default {
  issue: issueActions,
  timeEntry: timeEntryActions,
  settings: settingsActions
};
