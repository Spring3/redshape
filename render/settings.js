export const availableOptions = {
  // Used in OptionsBlock.jsx and settings.reducer.js
  // uiStyle, idleBehavior, progressSlider, timerCheckpoint
  uiStyle: [
    { value: 'default', label: 'Default' },
    { value: 'colors', label: 'Default with colors' },
    { value: 'enhanced', label: 'Enhanced' },
  ],
  idleBehavior: [
    { value: 'none', label: 'Do nothing' },
    { value: '1m', label: 'Pause if idle for 1 minute' },
    { value: '2m', label: 'Pause if idle for 2 minutes' },
    { value: '5m', label: 'Pause if idle for 5 minutes' },
    { value: '10m', label: 'Pause if idle for 10 minutes' },
    { value: '15m', label: 'Pause if idle for 15 minutes' },
    { value: '20m', label: 'Pause if idle for 20 minutes' },
    { value: '30m', label: 'Pause if idle for 30 minutes' },
    { value: '45m', label: 'Pause if idle for 45 minutes' },
    { value: '1h', label: 'Pause if idle for 1 hour' },
  ],
  progressSlider: [
    { value: '1%', label: 'Slider with 1% steps' },
    { value: '2%', label: 'Slider with 2% steps' },
    { value: '5%', label: 'Slider with 5% steps' },
    { value: '10%', label: 'Slider with 10% steps' },
  ],
  timerCheckpoint: [
    { value: 'none', label: 'Do not save state periodically' },
    { value: '1m', label: 'Save state every minute' },
    { value: '2m', label: 'Save state every 2 minutes' },
    { value: '3m', label: 'Save state every 3 minutes' },
    { value: '5m', label: 'Save state every 5 minutes' },
    { value: '10m', label: 'Save state every 10 minutes' },
    { value: '15m', label: 'Save state every 15 minutes' },
    { value: '30m', label: 'Save state every 30 minutes' },
  ],
  // Used in ColumnHeadersSelect.jsx and settings.reducer.js
  // issueHeaders
  issueHeaders: [
    {
      label: 'Id', isFixed: true, value: 'id', format: 'id'
    },
    { label: 'Project', value: 'project.name' },
    { label: 'Tracker', value: 'tracker.name', format: 'tracker' },
    { label: 'Status', value: 'status.name', format: 'status' },
    { label: 'Subject', isFixed: true, value: 'subject' },
    { label: 'Priority', value: 'priority.name', format: 'priority' },
    { label: 'Estimation', value: 'estimated_hours', format: 'hours' },
    { label: 'Total Estimation', value: 'total_estimated_hours', format: 'hours' },
    { label: 'Spent', value: 'spent_hours', format: 'hours' },
    { label: 'Total Spent', value: 'total_spent_hours', format: 'hours' },
    { label: 'Progress', value: 'done_ratio', format: 'progress' },
    { label: 'Tags', value: 'tags', format: 'tags' },
    { label: 'Due Date', value: 'due_date', format: 'date' },
    { label: 'Attachments', value: 'attachments', format: 'count' },
    { label: 'Assigned', value: 'assigned_to.name' },
    { label: 'Author', value: 'author.name' },
    { label: 'Updated', value: 'updated_on', format: 'date' },
    { label: 'Created', value: 'created_on', format: 'date' },
  ]
};
