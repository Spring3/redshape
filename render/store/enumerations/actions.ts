import type { IAction, Context } from 'overmind';
import { Activity, Response } from '../../../types';

const fetchActivities: IAction<void, Promise<Response<Activity[]>>> = async ({ effects, state }: Context) => {
  const response = await effects.mainProcess.request({
    payload: {
      method: 'GET',
      route: 'enumerations/time_entry_activities.json',
    }
  });

  if (response.success) {
    state.enumerations.activities = response.data.activities as Activity[] ?? [];
    return {
      success: true,
      data: state.enumerations.activities,
    };
  }

  return {
    success: false,
    error: response.error
  };
};

export {
  fetchActivities
};
