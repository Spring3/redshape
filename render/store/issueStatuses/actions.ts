import { IAction, Context } from 'overmind';
import { IssueStatus, PaginatedActionResponse } from '../../../types';
import { indexById } from '../helpers';

const getAll: IAction<void, Promise<PaginatedActionResponse<IssueStatus>>> = async ({ effects, state }: Context) => {
  const response = await effects.mainProcess.request({
    payload: {
      method: 'GET',
      route: 'issue_statuses.json'
    }
  });

  if (response.success) {
    state.issueStatuses.byId = {
      ...state.issueStatuses,
      ...indexById(response.payload.issueStatuses)
    };

    return {
      success: true,
      data: {
        items: response.payload.issueStatuses,
        total: response.payload.total,
        limit: response.payload.limit,
        offset: response.payload.offset
      },
      hasMore: response.payload.total > (response.payload.offset + response.payload.issueStatuses.length)
    };
  }

  return {
    success: false,
    data: {
      items: [],
      total: 0,
      limit: response.payload.limit,
      offset: response.payload.offset
    },
    hasMore: false,
    error: response.error
  };
};

export {
  getAll
};
