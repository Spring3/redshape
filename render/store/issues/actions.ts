import type { Context, IAction } from 'overmind';
import { Issue, PaginatedActionResponse, Response } from '../../../types';
import { indexById } from '../helpers';

type GetManyIssuesArgs = {
  filters: {
    assigned_to_id?: number | string;
    status_id?: '*' | 'open' | 'closed';
    subject?: string;
    sort?: string;
  },
  offset?: number;
  limit?: number;
}

const getMany: IAction<GetManyIssuesArgs, Promise<PaginatedActionResponse<Issue>>> = async ({ effects, state }: Context, { filters, offset = 0, limit = 20 }) => {
  const query = {
    limit: limit ? Math.abs(limit) : undefined,
    offset: offset ? Math.abs(offset) : 0,
    include: 'attachments,relations',
    ...filters
  };

  const response = await effects.mainProcess.request({
    payload: {
      route: 'issues.json',
      method: 'GET',
      query
    }
  });

  if (response.success) {
    state.issues.byId = {
      ...state.issues.byId,
      ...indexById(response.payload.issues)
    };

    console.log('payload', response.payload);

    return {
      success: true,
      data: {
        items: response.payload.issues,
        total: response.payload.total,
        limit: response.payload.limit,
        offset: response.payload.offset
      },
      hasMore: response.payload.total > (response.payload.offset + response.payload.issues.length),
    };
  }

  return {
    success: false,
    data: {
      items: [],
      total: 0,
      limit,
      offset
    },
    hasMore: false,
    error: response.error
  };
};

type GetOneIssueArgs = {
  id: string;
}

const getOne: IAction<GetOneIssueArgs, Promise<Response<Issue>>> = async ({ effects, state }: Context, { id }) => {
  const response = await effects.mainProcess.request({
    payload: {
      route: `issues/${id}.json`,
      method: 'GET',
      query: {
        include: 'attachments,children,relations,journals',
      }
    }
  });

  if (response.success) {
    const issue = response.payload as Issue;
    state.issues.byId[id] = issue;
    return {
      success: true,
      data: issue
    };
  }

  return {
    success: false,
    error: response.error
  };
};

export {
  getOne,
  getMany
};
