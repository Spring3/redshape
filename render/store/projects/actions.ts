import type { IAction, Context } from 'overmind';
import { PaginatedActionResponse } from '../../../types';
import { indexById } from '../helpers';

type GetManyProjectsArgs = {
  offset?: number;
  limit?: number;
}

const getManyProjects: IAction<GetManyProjectsArgs, Promise<PaginatedActionResponse<any>>> = async ({ state, effects }: Context, { offset = 0, limit = 20 }) => {
  const normalizedOffset = offset ? Math.abs(offset) : 0;
  const normalizedLimit = limit ? Math.abs(limit) : undefined;

  const response = await effects.mainProcess.request({
    payload: {
      route: 'projects.json',
      method: 'GET',
      query: {
        offset: normalizedOffset,
        include: 'time_entry_activities',
        limit: normalizedLimit
      }
    }
  });

  if (response.success) {
    state.projects.byId = {
      ...state.projects.byId,
      ...indexById(response.data.projects)
    };

    console.log('payload', response.data);

    return {
      success: true,
      data: {
        items: response.data.projects,
        total: response.data.total,
        limit: response.data.limit,
        offset: response.data.offset
      },
      hasMore: response.data.total > response.data.projects.length
    };
  }

  return {
    success: false,
    data: {
      items: [],
      total: 0,
      limit: response.data.limit,
      offset: response.data.offset
    },
    hasMore: false,
    error: response.error
  };
};

export {
  getManyProjects
};
