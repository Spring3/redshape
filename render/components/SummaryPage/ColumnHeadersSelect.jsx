import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import Select from 'react-select';
import { connect } from 'react-redux';
import makeAnimated from 'react-select/lib/animated';
import { withTheme } from 'styled-components';

import actions from '../../actions';
import { Label } from '../Input';

const selectStyles = {
  container: (base, state) => {
    return { ...base };
  },
  multiValue: (base, state) => {
    return state.data.isFixed
      ? { ...base, backgroundColor: '#FAFAFA', border: '1px solid #A4A4A4' }
      : { ...base, backgroundColor: 'transparent', border: '1px solid #3F3844' };
  },
  multiValueLabel: (base, state) => {
    return state.data.isFixed
      ? { ...base, paddingRight: 6, color: '#A4A4A4' }
      : base;
  },
  multiValueRemove: (base, state) => {
    return state.data.isFixed ? { ...base, display: 'none' } : base;
  }
};

class ColumnHeadersSelect extends Component {
  constructor(props) {
    super(props);

    this.issueHeaders = [
      { label: 'Id', isFixed: true, value: 'id' },
      { label: 'Project', value: 'project.name' },
      { label: 'Tracker', value: 'tracker.name', format: 'tracker' },
      { label: 'Status', value: 'status.name', format: 'status' },
      { label: 'Subject', isFixed: true, value: 'subject' },
      { label: 'Priority', value: 'priority.name', format: 'priority' },
      { label: 'Estimation', value: 'estimated_hours', format: 'hours' },
      { label: 'Total Estimation', value: 'total_estimated_hours', format: 'hours' },
      { label: 'Spent', value: 'spent_hours', format: 'hours' },
      { label: 'Total Spent', value: 'total_spent_hours', format: 'hours' },
      { label: 'Progress', value: 'done_ratio', format: 'progress' },
      { label: 'Due Date', value: 'due_date', format: 'date' }
    ];
  }

  onHeadersSelectChange = (value, { action, removedValue }) => {
    switch (action) {
      case 'remove-value':
      case 'pop-value':
        if (removedValue.isFixed) {
          return;
        }
        break;
    }

    this.props.settingsChangeIssueHeaders(value);
  }

  render() {
    const { theme, issueHeaders } = this.props;
    return (
      <Label htmlFor="headers" label="Table Columns">
        <Select
          name="headers"
          styles={selectStyles}
          components={makeAnimated()}
          options={this.issueHeaders}
          defaultValue={issueHeaders}
          value={issueHeaders}
          onChange={this.onHeadersSelectChange}
          isMulti={true}
          isClearable={false}
          theme={(defaultTheme) => ({
            ...defaultTheme,
            borderRadius: 3,
            colors: {
              ...defaultTheme.colors,
              primary: theme.main,
            },
          })
          }
        />
      </Label>
    );
  }
}

ColumnHeadersSelect.propTypes = {
  issueHeaders: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    format: PropTypes.string,
    isFixed: PropTypes.bool
  }).isRequired).isRequired,
  settingsChangeIssueHeaders: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  issueHeaders: state.settings.issueHeaders,
});

const mapDispatchToProps = dispatch => ({
  settingsChangeIssueHeaders: issueHeaders => dispatch(actions.settings.setIssueHeaders(issueHeaders))
});

export default withTheme(connect(mapStateToProps, mapDispatchToProps)(ColumnHeadersSelect));
