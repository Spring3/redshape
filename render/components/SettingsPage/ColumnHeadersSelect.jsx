import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import Select from 'react-select';
import { connect } from 'react-redux';
import makeAnimated from 'react-select/lib/animated';
import styled, { withTheme } from 'styled-components';
import HelpCircleIcon from 'mdi-react/HelpCircleIcon';
import { availableOptions } from '../../settings';

import actions from '../../actions';
import { Label } from '../Input';
import Tooltip from '../Tooltip';

const selectStyles = {
  container: (base, state) => ({ ...base }),
  multiValue: (base, state) => (state.data.isFixed
    ? { ...base, backgroundColor: '#FAFAFA', border: '1px solid #A4A4A4' }
    : { ...base, backgroundColor: 'transparent', border: '1px solid #3F3844' }),
  multiValueLabel: (base, state) => (state.data.isFixed
    ? { ...base, paddingRight: 6, color: '#A4A4A4' }
    : base),
  multiValueRemove: (base, state) => (state.data.isFixed ? { ...base, display: 'none' } : base)
};

const HelpIconStyled = styled(HelpCircleIcon)`
  padding-bottom: 1px;
`;
const LabelIcon = styled.span`
  margin-left: 0.2rem;
  color: #A0A0A0;
`;
const tooltipColumns = "'Tags', 'Total Estimation', 'Total Spent',\n'Spent', 'Subtasks' and colored 'Parent'\n need server-side support.";
const TableColumnsInfo = (<LabelIcon><Tooltip text={tooltipColumns}><HelpIconStyled size={14} /></Tooltip></LabelIcon>);

class ColumnHeadersSelect extends Component {
  constructor(props) {
    super(props);

    this.issueHeaders = availableOptions.issueHeaders;
  }

  onHeadersSelectChange = (value, { action, removedValue }) => {
    switch (action) {
      case 'remove-value':
      case 'pop-value':
        if (removedValue.isFixed) {
          return;
        }
        break;
    }

    this.props.settingsChangeIssueHeaders(value);
  }

  render() {
    const { theme, issueHeaders } = this.props;
    return (
      <Label htmlFor="headers" label="Table Columns" rightOfLabel={TableColumnsInfo}>
        <Select
          name="headers"
          styles={selectStyles}
          components={makeAnimated()}
          options={this.issueHeaders}
          defaultValue={issueHeaders}
          value={issueHeaders}
          onChange={this.onHeadersSelectChange}
          isMulti={true}
          isClearable={false}
          theme={defaultTheme => ({
            ...defaultTheme,
            borderRadius: 3,
            colors: {
              ...defaultTheme.colors,
              primary: theme.main,
            },
          })
          }
        />
      </Label>
    );
  }
}

ColumnHeadersSelect.propTypes = {
  issueHeaders: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    format: PropTypes.string,
    isFixed: PropTypes.bool
  }).isRequired).isRequired,
  settingsChangeIssueHeaders: PropTypes.func.isRequired,
  rightOfLabel: PropTypes.element,
};

const mapStateToProps = state => ({
  issueHeaders: state.settings.issueHeaders,
});

const mapDispatchToProps = dispatch => ({
  settingsChangeIssueHeaders: issueHeaders => dispatch(actions.settings.setIssueHeaders(issueHeaders))
});

export default withTheme(connect(mapStateToProps, mapDispatchToProps)(ColumnHeadersSelect));
