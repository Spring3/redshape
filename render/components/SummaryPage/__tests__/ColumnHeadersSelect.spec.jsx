import React from 'react';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { SETTINGS_ISSUE_HEADERS } from '../../../actions/settings.actions';
import ColumnHeadersSelect from '../ColumnHeadersSelect';
import theme from '../../../theme';

const mockStore = configureStore([thunk]);

describe('SummaryPage => ColumnHeadersSelect componnet', () => {
  it('should match the snapshot', () => {
    const state = {
      settings: {
        issueHeaders: [
          { label: 'Id', isFixed: true, value: 'id' },
          { label: 'Project', value: 'project.name' },
          { label: 'Tracker', value: 'tracker.name' },
          { label: 'Status', value: 'status.name' },
          { label: 'Subject', isFixed: true, value: 'subject' },
          { label: 'Priority', value: 'priority.name' },
          { label: 'Estimation', value: 'estimated_hours' },
          { label: 'Due Date', value: 'due_date' }
        ]
      }
    };
    const store = mockStore(state);
    const tree = renderer.create(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <ColumnHeadersSelect />
        </ThemeProvider>
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should emit an action when the issue headers are changed', () => {
    const state = {
      settings: {
        issueHeaders: [
          { label: 'Id', isFixed: true, value: 'id' },
          { label: 'Project', value: 'project.name' },
          { label: 'Tracker', value: 'tracker.name' },
          { label: 'Status', value: 'status.name' },
          { label: 'Subject', isFixed: true, value: 'subject' },
          { label: 'Priority', value: 'priority.name' },
          { label: 'Estimation', value: 'estimated_hours' },
          { label: 'Due Date', value: 'due_date' }
        ]
      },
      user: {
        id: 1,
        redmineEndpoint: 'https://redmine.redmine'
      }
    };
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <ColumnHeadersSelect />
        </ThemeProvider>
      </Provider>
    );
    const section = wrapper.find('ColumnHeadersSelect').instance();
    expect(section).toBeTruthy();

    section.onHeadersSelectChange('due_date', {});
    expect(store.getActions().length).toBe(1);
    expect(store.getActions()[0].type).toBe(SETTINGS_ISSUE_HEADERS);
  });
});
