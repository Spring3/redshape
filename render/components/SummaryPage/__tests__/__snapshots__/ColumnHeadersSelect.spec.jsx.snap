// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SummaryPage => ColumnHeadersSelect componnet should match the snapshot 1`] = `
.c0 h4 {
  margin-bottom: 10px;
  color: #A0A0A0;
}

.c1 {
  color: #A0A0A0;
}

.c6 {
  position: absolute;
  display: none;
  border-radius: 3px;
  top: -50px;
  left: 50%;
  width: auto;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  padding: 5px;
  background: #FFFFFF;
  color: #EF6069;
  text-align: center;
  font-size: 12px;
  white-space: nowrap;
  box-shadow: 0px 2px 7px #A0A0A0;
}

.c6::after {
  content: ' ';
  position: absolute;
  top: 100%;
  left: 50%;
  width: auto;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  border: 2px solid black;
  border-width: 5px;
  border-color: #FFFFFF transparent transparent transparent;
}

.c3 {
  display: inline-block;
  position: relative;
}

.c3:hover .c5 {
  display: block;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  padding-bottom: 1px;
}

.c2 {
  margin-left: 0.2rem;
  color: #A0A0A0;
}

<div
  className="form-group undefined c0"
>
  <h4
    className="c1"
    htmlFor="headers"
  >
    Table Columns
    <span
      className="c2"
    >
      <div
        className="c3"
      >
        <svg
          className="mdi-icon c4"
          fill="currentColor"
          height={14}
          viewBox="0 0 24 24"
          width={14}
        >
          <path
            d="M15.07,11.25L14.17,12.17C13.45,12.89 13,13.5 13,15H11V14.5C11,13.39 11.45,12.39 12.17,11.67L13.41,10.41C13.78,10.05 14,9.55 14,9C14,7.89 13.1,7 12,7A2,2 0 0,0 10,9H8A4,4 0 0,1 12,5A4,4 0 0,1 16,9C16,9.88 15.64,10.67 15.07,11.25M13,19H11V17H13M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12C22,6.47 17.5,2 12,2Z"
          />
        </svg>
        <p
          className="tooltip c5 c6"
        >
          <div
            className="c7"
          >
            <span>
              'Tags', 'Total Estimation', 'Total Spent'
            </span>
            <span>
              and 'Spent' need server-side support.
            </span>
          </div>
        </p>
      </div>
    </span>
  </h4>
  <div
    className="css-10nd86i"
    onKeyDown={[Function]}
  >
    <div
      className="css-1rcfx4g"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="css-1hwfws3"
      >
        <div
          style={
            Object {
              "overflow": "hidden",
              "whiteSpace": "nowrap",
              "width": "auto",
            }
          }
        >
          <div
            className="css-iuzuh3"
          >
            <div
              className="css-7ydjiw"
            >
              Id
            </div>
            <div
              className="css-azd3r6"
              onClick={[Function]}
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-19bqh2r"
                focusable="false"
                height={14}
                viewBox="0 0 20 20"
                width={14}
              >
                <path
                  d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "overflow": "hidden",
              "whiteSpace": "nowrap",
              "width": "auto",
            }
          }
        >
          <div
            className="css-1gneeqd"
          >
            <div
              className="css-4a5pmt"
            >
              Project
            </div>
            <div
              className="css-sn3edf"
              onClick={[Function]}
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-19bqh2r"
                focusable="false"
                height={14}
                viewBox="0 0 20 20"
                width={14}
              >
                <path
                  d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "overflow": "hidden",
              "whiteSpace": "nowrap",
              "width": "auto",
            }
          }
        >
          <div
            className="css-1gneeqd"
          >
            <div
              className="css-4a5pmt"
            >
              Tracker
            </div>
            <div
              className="css-sn3edf"
              onClick={[Function]}
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-19bqh2r"
                focusable="false"
                height={14}
                viewBox="0 0 20 20"
                width={14}
              >
                <path
                  d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "overflow": "hidden",
              "whiteSpace": "nowrap",
              "width": "auto",
            }
          }
        >
          <div
            className="css-1gneeqd"
          >
            <div
              className="css-4a5pmt"
            >
              Status
            </div>
            <div
              className="css-sn3edf"
              onClick={[Function]}
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-19bqh2r"
                focusable="false"
                height={14}
                viewBox="0 0 20 20"
                width={14}
              >
                <path
                  d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "overflow": "hidden",
              "whiteSpace": "nowrap",
              "width": "auto",
            }
          }
        >
          <div
            className="css-iuzuh3"
          >
            <div
              className="css-7ydjiw"
            >
              Subject
            </div>
            <div
              className="css-azd3r6"
              onClick={[Function]}
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-19bqh2r"
                focusable="false"
                height={14}
                viewBox="0 0 20 20"
                width={14}
              >
                <path
                  d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "overflow": "hidden",
              "whiteSpace": "nowrap",
              "width": "auto",
            }
          }
        >
          <div
            className="css-1gneeqd"
          >
            <div
              className="css-4a5pmt"
            >
              Priority
            </div>
            <div
              className="css-sn3edf"
              onClick={[Function]}
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-19bqh2r"
                focusable="false"
                height={14}
                viewBox="0 0 20 20"
                width={14}
              >
                <path
                  d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "overflow": "hidden",
              "whiteSpace": "nowrap",
              "width": "auto",
            }
          }
        >
          <div
            className="css-1gneeqd"
          >
            <div
              className="css-4a5pmt"
            >
              Estimation
            </div>
            <div
              className="css-sn3edf"
              onClick={[Function]}
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-19bqh2r"
                focusable="false"
                height={14}
                viewBox="0 0 20 20"
                width={14}
              >
                <path
                  d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "overflow": "hidden",
              "whiteSpace": "nowrap",
              "width": "auto",
            }
          }
        >
          <div
            className="css-1gneeqd"
          >
            <div
              className="css-4a5pmt"
            >
              Due Date
            </div>
            <div
              className="css-sn3edf"
              onClick={[Function]}
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-19bqh2r"
                focusable="false"
                height={14}
                viewBox="0 0 20 20"
                width={14}
              >
                <path
                  d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          className="css-1g6gooi"
        >
          <div
            className=""
            style={
              Object {
                "display": "inline-block",
              }
            }
          >
            <input
              aria-autocomplete="list"
              autoCapitalize="none"
              autoComplete="off"
              autoCorrect="off"
              disabled={false}
              id="react-select-2-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              spellCheck="false"
              style={
                Object {
                  "background": 0,
                  "border": 0,
                  "boxSizing": "content-box",
                  "color": "inherit",
                  "fontSize": "inherit",
                  "opacity": 1,
                  "outline": 0,
                  "padding": 0,
                  "width": "1px",
                }
              }
              tabIndex="0"
              type="text"
              value=""
            />
            <div
              style={
                Object {
                  "height": 0,
                  "left": 0,
                  "overflow": "scroll",
                  "position": "absolute",
                  "top": 0,
                  "visibility": "hidden",
                  "whiteSpace": "pre",
                }
              }
            >
              
            </div>
          </div>
        </div>
      </div>
      <div
        className="css-1wy0on6"
      >
        <span
          className="css-d8oujb"
        />
        <div
          aria-hidden="true"
          className="css-1ep9fjw"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <svg
            aria-hidden="true"
            className="css-19bqh2r"
            focusable="false"
            height={20}
            viewBox="0 0 20 20"
            width={20}
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div>
      <input
        name="headers"
        type="hidden"
        value="id"
      />
      <input
        name="headers"
        type="hidden"
        value="project.name"
      />
      <input
        name="headers"
        type="hidden"
        value="tracker.name"
      />
      <input
        name="headers"
        type="hidden"
        value="status.name"
      />
      <input
        name="headers"
        type="hidden"
        value="subject"
      />
      <input
        name="headers"
        type="hidden"
        value="priority.name"
      />
      <input
        name="headers"
        type="hidden"
        value="estimated_hours"
      />
      <input
        name="headers"
        type="hidden"
        value="due_date"
      />
    </div>
  </div>
</div>
`;
